File: shortest-word-2.js
// Time, Space - O(N), O(N)

/**
 * @param {string[]} words
 */

let map;
var WordDistance = function(words) {
    map = {}
    
    for(let i=0;i<words.length;i++) {
        let word = words[i];
        map = {
            ...map,
            [word]: map[word] ? [...map[word], i] : [i]
        }
    }
    
};

/** 
 * @param {string} word1 
 * @param {string} word2
 * @return {number}
 */
WordDistance.prototype.shortest = function(word1, word2) {
    
    let p1 = 0;
    let p2 = 0;
    let min = Number.MAX_VALUE;
    while(p1 < map[word1].length && p2 < map[word2].length) {
         
        let index1 = map[word1][p1];
        let index2 = map[word2][p2];
        
        if(index1 < index2) {
            p1++;
        }
        else {
            p2++;
        }
        
        min = Math.min(min, Math.abs(index1 - index2));
    } 
    
    return min;
};

/** 
 * Your WordDistance object will be instantiated and called as such:
 * var obj = new WordDistance(words)
 * var param_1 = obj.shortest(word1,word2)
 */

File: shortestWord2.java
// Time Complexity :O(m+k)
// Space Complexity :constant 
class WordDistance {
    private HashMap<String, List<Integer>> map;

    public WordDistance(String[] wordsDict) {
        map = new HashMap<>();
        // create map to record indices of words so that we dont have to go through
        // whole list each time
        for (int i = 0; i < wordsDict.length; i++) {
            String curr = wordsDict[i];
            if (!map.containsKey(curr)) {
                map.put(curr, new ArrayList<>());
            }
            map.get(curr).add(i);
        }
    }

    public int shortest(String word1, String word2) {
        List<Integer> li1 = map.get(word1);
        List<Integer> li2 = map.get(word2);

        int p1 = 0;
        int p2 = 0;
        int min = Integer.MAX_VALUE;
        // get both lists and find minimum distance between the words
        while (p1 < li1.size() && p2 < li2.size()) {
            int idx1 = li1.get(p1);
            int idx2 = li2.get(p2);
            min = Math.min(min, Math.abs(idx2 - idx1));
            if (idx1 < idx2) {
                p1++;
            } else {
                p2++;
            }
        }
        return min;
    }
}

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance obj = new WordDistance(wordsDict);
 * int param_1 = obj.shortest(word1,word2);
 */